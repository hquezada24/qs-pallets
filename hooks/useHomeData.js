import { useState, useEffect } from "react";

export const useHomeData = () => {
  const [data, setData] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  const API_BASE_URL =
    import.meta.env.VITE_API_BASE_URL || "http://localhost:3000";

  useEffect(() => {
    const fetchData = async () => {
      try {
        const response = await fetch(API_BASE_URL);
        if (!response.ok) throw new Error("Failed to fetch");
        const result = await response.json();
        setData(result);
      } catch (err) {
        setError(err.message);
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, []);

  return { data, loading, error };
};
