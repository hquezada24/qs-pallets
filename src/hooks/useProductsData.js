import { useState, useEffect } from "react";

export const useProductsData = () => {
  const [data, setData] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    const API_BASE_URL =
      import.meta.env.VITE_API_PRODUCTS || "http://localhost:3000/products";
    const fetchData = async () => {
      try {
        const response = await fetch(API_BASE_URL);
        if (!response.ok) throw new Error("Failed to fetch");
        const result = await response.json();
        setData(result);
      } catch (err) {
        setError(err.message);
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, []);

  return { data, loading, error };
};
