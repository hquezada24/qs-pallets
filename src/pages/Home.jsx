import styles from "./Styles.module.css";
import { Hero } from "./components/Hero/Hero";

const Home = () => {
  return (
    <div className={styles.home}>
      <Hero />
    </div>
  );
};

export { Home };
