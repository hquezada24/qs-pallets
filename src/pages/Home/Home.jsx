import styles from "./Styles.module.css";
import { Hero } from "./Components/Hero/Hero";
import { ScrollRestoration } from "react-router-dom";
import { FeaturedProducts } from "./Components/FeaturedProducts/FeaturedProducts";
import { WhyChooseUs } from "./Components/WhyChooseUs/WhyChooseUs";
import { CallToAction } from "./Components/CallToAction/CallToAction";

const Home = () => {
  return (
    <div className={styles.home}>
      <ScrollRestoration />
      <Hero />
      <FeaturedProducts />
      <WhyChooseUs />
      <CallToAction />
    </div>
  );
};

export { Home };
