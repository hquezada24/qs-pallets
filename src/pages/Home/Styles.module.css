/* Main Home Container */
.home {
  width: 100%;
  min-height: 100vh;
  background: linear-gradient(135deg, #f8fff8 0%, #ffffff 100%);
  position: relative;
  overflow-x: hidden;
}

/* Animated Background Effects */
.background {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: linear-gradient(135deg, #228b22 0%, #32cd32 50%, #90ee90 100%);
  z-index: 1;
}

.background::before {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: url("data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 200 200'><defs><pattern id='pallets' width='40' height='40' patternUnits='userSpaceOnUse'><rect width='40' height='40' fill='none'/><rect x='5' y='5' width='30' height='6' fill='rgba(255,255,255,0.1)' rx='1'/><rect x='5' y='15' width='30' height='6' fill='rgba(255,255,255,0.1)' rx='1'/><rect x='5' y='25' width='30' height='6' fill='rgba(255,255,255,0.1)' rx='1'/><rect x='10' y='0' width='4' height='40' fill='rgba(255,255,255,0.08)'/><rect x='26' y='0' width='4' height='40' fill='rgba(255,255,255,0.08)'/></pattern></defs><rect width='200' height='200' fill='url(%23pallets)'/></svg>")
    repeat;
  opacity: 0.3;
  animation: floatPattern 30s linear infinite;
}

.background::after {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: radial-gradient(
      circle at 30% 70%,
      rgba(255, 255, 255, 0.1) 0%,
      transparent 50%
    ),
    radial-gradient(
      circle at 80% 20%,
      rgba(255, 255, 255, 0.15) 0%,
      transparent 50%
    );
}

.error {
  height: 80vh;
  display: flex;
  align-items: center;
  justify-content: center;
}

@keyframes floatPattern {
  0% {
    transform: translate(0, 0) rotate(0deg);
  }
  25% {
    transform: translate(-10px, -10px) rotate(90deg);
  }
  50% {
    transform: translate(-20px, 0px) rotate(180deg);
  }
  75% {
    transform: translate(-10px, 10px) rotate(270deg);
  }
  100% {
    transform: translate(0, 0) rotate(360deg);
  }
}
