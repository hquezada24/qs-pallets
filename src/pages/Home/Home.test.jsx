// Home.test.jsx
// @vitest-environment jsdom
import { describe, it, expect, beforeEach, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { Home } from "./Home";
import * as useHomeDataModule from "../../hooks/useHomeData";

// Mock the custom hook module
vi.mock("../../hooks/useHomeData");

// Mock ScrollRestoration from react-router-dom
vi.mock("react-router-dom", async () => {
  const actual = await vi.importActual("react-router-dom");
  return {
    ...actual,
    ScrollRestoration: () => null,
  };
});

// Mock child components
vi.mock("./Components/Hero/Hero", () => ({
  Hero: () => <div data-testid="hero">Hero Component</div>,
}));

vi.mock("./Components/FeaturedProducts/FeaturedProducts", () => ({
  FeaturedProducts: ({ products }) => (
    <div data-testid="featured-products">
      Featured Products: {products?.length || 0}
    </div>
  ),
}));

vi.mock("./Components/WhyChooseUs/WhyChooseUs", () => ({
  WhyChooseUs: ({ features }) => (
    <div data-testid="why-choose-us">
      Why Choose Us: {features?.length || 0}
    </div>
  ),
}));

vi.mock("./Components/CallToAction/CallToAction", () => ({
  CallToAction: () => <div data-testid="cta">Call to Action</div>,
}));

// Wrapper component for Router
const renderWithRouter = (component) => {
  return render(<BrowserRouter>{component}</BrowserRouter>);
};

describe("Home Component", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders loading state", () => {
    vi.spyOn(useHomeDataModule, "useHomeData").mockReturnValue({
      data: null,
      loading: true,
      error: null,
    });

    renderWithRouter(<Home />);

    expect(screen.getByText("Loading...")).toBeInTheDocument();
    expect(screen.queryByTestId("hero")).not.toBeInTheDocument();
  });

  it("renders error state", () => {
    const errorMessage = "Failed to fetch data";
    vi.spyOn(useHomeDataModule, "useHomeData").mockReturnValue({
      data: null,
      loading: false,
      error: errorMessage,
    });

    renderWithRouter(<Home />);

    expect(screen.getByText(`Error: ${errorMessage}`)).toBeInTheDocument();
    expect(screen.queryByTestId("hero")).not.toBeInTheDocument();
  });

  it("renders all components when data is loaded successfully", () => {
    const mockData = {
      features: [{ id: 1, name: "Product 1" }],
      whyChooseUs: [{ id: 1, title: "Feature 1" }],
    };

    vi.spyOn(useHomeDataModule, "useHomeData").mockReturnValue({
      data: mockData,
      loading: false,
      error: null,
    });

    const { container } = renderWithRouter(<Home />);

    expect(screen.getByTestId("hero")).toBeInTheDocument();
    expect(screen.getByTestId("featured-products")).toBeInTheDocument();
    expect(screen.getByTestId("why-choose-us")).toBeInTheDocument();
    expect(screen.getByTestId("cta")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("passes correct props to FeaturedProducts", () => {
    const mockProducts = [
      { id: 1, name: "Product 1" },
      { id: 2, name: "Product 2" },
    ];

    vi.spyOn(useHomeDataModule, "useHomeData").mockReturnValue({
      data: { features: mockProducts, whyChooseUs: [] },
      loading: false,
      error: null,
    });

    renderWithRouter(<Home />);

    expect(screen.getByText("Featured Products: 2")).toBeInTheDocument();
  });

  it("passes correct props to WhyChooseUs", () => {
    const mockFeatures = [
      { id: 1, title: "Feature 1" },
      { id: 2, title: "Feature 2" },
      { id: 3, title: "Feature 3" },
    ];

    vi.spyOn(useHomeDataModule, "useHomeData").mockReturnValue({
      data: { features: [], whyChooseUs: mockFeatures },
      loading: false,
      error: null,
    });

    renderWithRouter(<Home />);

    expect(screen.getByText("Why Choose Us: 3")).toBeInTheDocument();
  });

  it("handles undefined data gracefully", () => {
    vi.spyOn(useHomeDataModule, "useHomeData").mockReturnValue({
      data: undefined,
      loading: false,
      error: null,
    });

    renderWithRouter(<Home />);

    expect(screen.getByTestId("hero")).toBeInTheDocument();
    expect(screen.getByTestId("featured-products")).toBeInTheDocument();
    expect(screen.getByText("Featured Products: 0")).toBeInTheDocument();
  });

  it("renders successfully", () => {
    const { container } = renderWithRouter(<Home />);
    expect(container.firstChild).toBeInTheDocument();
  });
});
