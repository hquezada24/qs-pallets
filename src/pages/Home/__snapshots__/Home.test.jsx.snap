// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home Component > renders all components when data is loaded successfully 1`] = `
<div>
  <div
    class="_home_0da9fa"
  >
    <div
      data-testid="hero"
    >
      Hero Component
    </div>
    <div
      data-testid="featured-products"
    >
      Featured Products: 
      1
    </div>
    <div
      data-testid="why-choose-us"
    >
      Why Choose Us: 
      1
    </div>
    <div
      data-testid="cta"
    >
      Call to Action
    </div>
  </div>
</div>
`;
